libdrgn/python/thread.c

```diff
@@ -0,0 +1,113 @@
+#include "drgnpy.h"
+
+static void Thread_dealloc(Thread *self)
+{
+	drgn_thread_destroy(&self->thread);
+	Py_XDECREF(self->prog);
+	Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static PyObject *Thread_get_tid(Thread *self)
+{
+	return PyLong_FromUnsignedLong(self->thread.tid);
+}
+
+static PyObject *Thread_get_prstatus(Thread *self)
+{
+	return PyBytes_FromStringAndSize(self->thread.prstatus.str,
+					 self->thread.prstatus.len);
+}
+
+static PyObject *Thread_get_object(Thread *self)
+{
+	if (!(self->prog->prog.flags & DRGN_PROGRAM_IS_LINUX_KERNEL)) {
+		PyErr_SetString(
+			PyExc_AttributeError,
+			"a Thread's associated object is only defined when debugging the kernel");
+		return NULL;
+	}
+	DrgnObject *object = DrgnObject_alloc(self->prog);
+	if (!object)
+		return NULL;
+	object->obj = self->thread.thread;
+	return (PyObject *)object;
+}
+
+static PyObject *Thread_stack_trace(Thread *self)
+{
+	struct drgn_error *err;
+	struct drgn_stack_trace *trace;
+	err = drgn_program_stack_trace(&self->prog->prog, self->thread.tid,
+				       &trace);
```

Comment:
I think we also want a `struct drgn_error *drgn_thread_stack_trace(struct drgn_thread *thread, struct drgn_stack_trace **ret)` rather than doing it by TID:
* For userspace core dumps, `struct drgn_thread` already has the `PRSTATUS` note, so `drgn_get_initial_registers()` shouldn't need to look it up again.
* For kernel core dumps, we can cache the `PRSTATUS` note for `crashed_thread` and avoid redoing the `PRSTATUS` lookup. (You won't be able to keep `prstatus` and the `drgn_object` in a union anymore.)
* For the live kernel and threads in kernel core dumps other than `crashed_thread`, `struct drgn_thread` already has the `task_struct` object, so `drgn_get_initial_registers` shouldn't need to look it up again.

Reply:
For the live kernel case, wouldn't you actually want to look up the `task_struct` again inside `drgn_get_initial_registers`, since it's possible that the `task_struct` for the TID you're interested in no longer exists, or the TID has been recycled?

Reply:
Doing this with the thread API:

```python
thread = prog.thread(123)
thread.stack_trace()
```

Is more or less the same as doing this with the existing APIs:

```python
task = find_task(prog, 123)
prog.stack_trace(task)
```

There's always an implicit race for the live kernel, so I don't think we need to look it up again.

Reply:
Alright. Will likely have to modify the internals of several of the existing functions or extract some of their contents into static functions, so this might take a bit.